martin/src/cog/image.rs

```diff
@@ -0,0 +1,234 @@
+
+use super::CogError;
+use crate::{MartinResult, TileData};
+
+/// Image represents a single image in a COG file. A tiff file may contain many images.
+/// This type contains several useful information and method for taking tiles from the image.
+#[derive(Clone, Debug)]
+pub struct Image {
+    /// IFD(Image file directory) number.
+    /// An IFD contains information about the image, as well as pointers to the actual image data.
+    pub ifd: usize,
+    /// How many tiles in a row of this image
+    pub across: u32,
+    // How many tiles in a column of this image
+    pub down: u32,
+}
+
+impl Image {
+    #[allow(clippy::cast_sign_loss)]
+    #[allow(clippy::cast_possible_truncation)]
+    pub fn get_tile(
+        &self,
+        decoder: &mut Decoder<File>,
+        xyz: TileCoord,
+        nodata: Option<f64>,
+        path: &Path,
+    ) -> MartinResult<TileData> {
+        decoder
+            .seek_to_image(self.ifd)
+            .map_err(|e| CogError::IfdSeekFailed(e, self.ifd, path.to_path_buf()))?;
+
+        let tile_idx;
+        if let Some(idx) = self.get_tile_idx(xyz) {
+            tile_idx = idx;
+        } else {
+            return Ok(Vec::new());
+        }
+        let decode_result = decoder
+            .read_chunk(tile_idx)
+            .map_err(|e| CogError::ReadChunkFailed(e, tile_idx, self.ifd, path.to_path_buf()))?;
+        let color_type = decoder
+            .colortype()
+            .map_err(|e| CogError::InvalidTiffFile(e, path.to_path_buf()))?;
+
+        let (tile_width, tile_height) = decoder.chunk_dimensions();
+        let (data_width, data_height) = decoder.chunk_data_dimensions(tile_idx);
+
+        //do more research on the not u8 case, is this the right way to do it?
+        let png_file_bytes = match (decode_result, color_type) {
+            (DecodingResult::U8(vec), tiff::ColorType::RGB(_)) => rgb_to_png(
+                vec,
+                (tile_width, tile_height),
+                (data_width, data_height),
+                3,
+                nodata.map(|v| v as u8),
+                path,
+            ),
+            (DecodingResult::U8(vec), tiff::ColorType::RGBA(_)) => rgb_to_png(
+                vec,
+                (tile_width, tile_height),
+                (data_width, data_height),
+                4,
+                nodata.map(|v| v as u8),
+                path,
+            ),
+            (_, _) => Err(CogError::NotSupportedColorTypeAndBitDepth(
+                color_type,
+                path.to_path_buf(),
+            )),
+            // do others in next PRs, a lot of discussion would be needed
+        }?;
+        Ok(png_file_bytes)
+    }
+    fn get_tile_idx(&self, xyz: TileCoord) -> Option<u32> {
+        let across = self.across;
+        let down = self.down;
+        if xyz.y >= down || xyz.x >= across {
+            return None;
+        }
+
+        let tile_idx = xyz.y * across + xyz.x;
+        if tile_idx >= across * down {
```

Comment:
nit: For things like this, I think this is more readable.

Feel free to dismiss if you think otherwise

```suggestion
        if xyz.y >= self.down || xyz.x >= self.across {
            return None;
        }

        let tile_idx = xyz.y * self.across + xyz.x;
        if tile_idx >= self.across * self.down {
```